crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -714,22 +716,42 @@ impl PublishFlakeRef {
         flox: &Flox,
         git_service_prefer_https: bool,
     ) -> Result<Self, ConvertFlakeRefError> {
-        let publish_flake_ref = match flake_ref {
-            FlakeRef::GitSsh(ssh_ref) => Self::Ssh(ssh_ref),
-            FlakeRef::GitHttps(https_ref) => Self::Https(https_ref),
-            // resolve upstream for local git repo
-            FlakeRef::GitPath(file_ref) => {
-                Self::from_git_file_flake_ref(file_ref, &flox.nix(Default::default())).await?
-            },
-            // resolve indirect ref to direct ref (recursively)
-            FlakeRef::Indirect(_) => todo!(),
-            FlakeRef::Github(github_ref) => {
-                Self::from_github_ref(github_ref, git_service_prefer_https)?
-            },
-            FlakeRef::Gitlab(_) => todo!(),
-            _ => Err(ConvertFlakeRefError::UnsupportedTarget(flake_ref))?,
+        // This should really be a recursive function, but that requires two things:
+        // - Making this return type Pin<Box<impl Future<Output = ...>>>
+        // - That all Futures that this one depends on are Send
+        // It turns out we can't do that because the GitProvider trait doesn't require the Futures it
+        // returns to be Send. Instead of a recursive call we do this loop until we no longer have
+        // an indirect flake reference.
+        let mut flake_ref = flake_ref;
+        let publish_flakeref = loop {
+            match flake_ref.clone() {
+                FlakeRef::GitSsh(ssh_ref) => {
+                    break Self::Ssh(ssh_ref);
+                },
+                FlakeRef::GitHttps(https_ref) => {
+                    break Self::Https(https_ref);
+                },
+                // resolve upstream for local git repo
+                FlakeRef::GitPath(file_ref) => {
+                    break Self::from_git_file_flake_ref(file_ref, &flox.nix(Default::default()))
+                        .await?;
+                },
+                // resolve indirect ref to direct ref (recursively)
+                FlakeRef::Indirect(indirect) => match indirect.resolve()? {
+                    FlakeRef::Indirect(_) => unreachable!(),
```

Comment:
How would this loop if `FlakeRef::Indirect(_)` is unreachable?

Reply:
The loop executes a maximum of twice, it's just more concise to express as a loop rather than two match expressions. If you resolve an indirect reference you shouldn't get back another indirect reference.

1 iteration: you call this with something other than an indirect reference.
2 iterations: you call it with an indirect reference, first iteration is resolving to a GitHub reference (or something else), second iteration is resolving GitHub to Git+File (or SSH, etc).